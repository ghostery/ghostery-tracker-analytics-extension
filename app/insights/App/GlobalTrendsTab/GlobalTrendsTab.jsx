import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import isEqual from 'lodash.isequal';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Tooltip from 'react-bootstrap/Tooltip';

import GlobalTrendsListItem from '../GlobalTrendsListItem';
import { convertBytes } from '../../../../src/utils/convert';
import './GlobalTrendsTab.scss';

class GlobalTrendsTab extends Component {
  constructor(props) {
    super(props);
    this.state = { sort: 'favorites' };
  }

  componentDidMount() {
    this.fetchNewWTMData(true);
  }

  componentDidUpdate(prevProps) {
    this.fetchNewWTMData(false, prevProps);
  }

  fetchNewWTMData = (initialFetch, prevProps) => {
    const { actions, trackerInformation, wtmData } = this.props;
    if (initialFetch || !(isEqual(prevProps.trackerInformation, trackerInformation))) {
      const newTrackers = [];
      Object.keys(trackerInformation).forEach((trackerId) => {
        if (!(trackerId in wtmData)) {
          newTrackers.push(trackerId);
        }
      });
      actions.fetchWTMData(newTrackers);
    }
  }

  handleStar = (trackerId, trackerName) => {
    const { starredTrackerIds, messageCreators } = this.props;

    if (starredTrackerIds.hasOwnProperty(trackerId)) {
      messageCreators.removeStarredTracker({ trackerId, trackerName, view: '5' });
    } else {
      messageCreators.addStarredTracker({ trackerId, trackerName, view: '5' });
    }
  }

  handleClick = (sortType) => {
    const { messageCreators } = this.props;
    const sortTypeMapping = {
      favorites: '0',
      prevalence: '1',
      category: '2',
    };
    messageCreators.sendMetrics({
      type: 'sort_trackerInformationTab',
      sortType: sortTypeMapping[sortType],
    });
    this.setState({ sort: sortType });
  }

  sortTrackerIds = () => {
    const { sort } = this.state;
    const { trackerInformation, wtmData, starredTrackerIds } = this.props;

    const sortedTrackerIds = Object.keys(trackerInformation);
    sortedTrackerIds.sort((a, b) => (
      trackerInformation[a].name.toLowerCase().localeCompare(
        trackerInformation[b].name.toLowerCase(),
      )
    ));

    if (sort === 'favorites') {
      sortedTrackerIds.sort((a, b) => {
        if ((a in starredTrackerIds) === (b in starredTrackerIds)) { return 0; }
        if (a in starredTrackerIds) { return -1; }
        return 1;
      });
    } else if (sort === 'prevalence') {
      sortedTrackerIds.sort((a, b) => {
        if (!wtmData[a] && !wtmData[b]) { return 0; }
        if (wtmData[a] && !wtmData[b]) { return -1; }
        if (!wtmData[a] && wtmData[b]) { return 1; }
        return wtmData[a].overview.reach_rank - wtmData[b].overview.reach_rank;
      });
    } else if (sort === 'category') {
      sortedTrackerIds.sort((a, b) => (
        trackerInformation[a].category.toLowerCase()
          .localeCompare(trackerInformation[b].category.toLowerCase())
      ));
    }

    return sortedTrackerIds;
  }

  render() {
    const { sort } = this.state;
    const { wtmData, trackerInformation, starredTrackerIds, totalSize } = this.props;
    const setButtonClasses = sortType => (
      ClassNames(
        'GlobalTrendsTab__headerButton',
        'd-flex',
        'justify-content-center',
        'align-items-center',
        { active: sortType },
      )
    );

    const sortedTrackerIds = this.sortTrackerIds();

    const favoritesClassNames = setButtonClasses(sort === 'favorites');
    const prevalenceClassNames = setButtonClasses(sort === 'prevalence');
    const categoryClassNames = setButtonClasses(sort === 'category');

    return (
      <div className="GlobalTrendsTab d-flex justify-content-between align-items-start">
        <div className="GlobalTrendsTab__list GlobalTrendsTab--opaqueBackground">
          <div className="GlobalTrendsTab__listHeader d-flex flex-column justify-content-between">
            <div className="d-flex justify-content-start align-items-center">
              <h3 className="GlobalTrendsTab--marginRight">
                Known Tracker Trends
              </h3>
              <OverlayTrigger
                placement="auto"
                overlay={(
                  <Tooltip className="tooltipRight">
                    View available statistics and trends over time for trackers on this page
                  </Tooltip>
                )}
              >
                <img
                  className="infoIcon"
                  src="/dist/images/shared/info-icon.svg"
                  width="16"
                  height="16"
                  alt="More info hover: View known tracker trends"
                />
              </OverlayTrigger>
            </div>
            <div className="GlobalTrendsTab__headerButtonList d-flex justify-content-between align-items-center">
              <p>Sort by: </p>
              <OverlayTrigger
                placement="top"
                delay={{ show: 500 }}
                overlay={(
                  <Tooltip className="tooltipLeft">
                    Sort by favorited trackers
                  </Tooltip>
                )}
              >
                <div
                  className={favoritesClassNames}
                  onClick={() => this.handleClick('favorites')}
                >
                  <p>Favorites</p>
                </div>
              </OverlayTrigger>
              <OverlayTrigger
                placement="top"
                delay={{ show: 500 }}
                overlay={(
                  <Tooltip className="tooltipLeft">
                    Sort by most common trackers
                  </Tooltip>
                )}
              >
                <div
                  className={prevalenceClassNames}
                  onClick={() => this.handleClick('prevalence')}
                >
                  <p>Prevalence</p>
                </div>
              </OverlayTrigger>
              <OverlayTrigger
                placement="top"
                delay={{ show: 500 }}
                overlay={(
                  <Tooltip className="tooltipLeft">Sort by tracker type</Tooltip>
                )}
              >
                <div
                  className={categoryClassNames}
                  onClick={() => this.handleClick('category')}
                >
                  <p>Category</p>
                </div>
              </OverlayTrigger>
            </div>
          </div>
          <div className="GlobalTrendsTab__tableHeader">
            <Container fluid className="GlobalTrendsTab__tableHeaderContainer">
              <Row>
                <Col className="GlobalTrendsListItem--paddingZero" sm={1} />
                <Col className="GlobalTrendsListItem__prevalence GlobalTrendsListItem--paddingZero" sm={2}>Prevalence</Col>
                <Col className="GlobalTrendsListItem--paddingZero" sm={5}>Tracker</Col>
                <Col className="GlobalTrendsListItem--paddingZero" sm={4}>Category</Col>
              </Row>
            </Container>
          </div>
          {sortedTrackerIds.map(trackerId => (
            <GlobalTrendsListItem
              key={trackerId}
              id={trackerId}
              name={trackerInformation[trackerId].name}
              category={trackerInformation[trackerId].category}
              listItemWtmData={wtmData[trackerId] || false}
              handleStar={this.handleStar}
              starred={starredTrackerIds.hasOwnProperty(trackerId)}
            />
          ))}
        </div>
        <div className="GlobalTrendsTab__stats">
          <div className="GlobalTrendsTab__statDisplay GlobalTrendsTab--opaqueBackground">
            <div className="d-flex justify-content-start align-items-center">
              <h3 className="GlobalTrendsTab--marginRight">
                Total Tracker Size on this Page
              </h3>
              <OverlayTrigger
                placement="auto"
                overlay={(
                  <Tooltip className="tooltipLeft">
                    Aggregate size of all trackers found on this page
                  </Tooltip>
                )}
              >
                <img
                  src="/dist/images/shared/info-icon.svg"
                  className="infoIcon"
                  width="16"
                  height="16"
                  alt="More info hover: Aggregate size of all trackers found on this page"
                />
              </OverlayTrigger>
            </div>
            <p className="GlobalTrendsTab__statText">{convertBytes(totalSize)}</p>
          </div>
        </div>
      </div>
    );
  }
}

GlobalTrendsTab.propTypes = {
  actions: PropTypes.shape({
    fetchWTMData: PropTypes.func.isRequired,
  }).isRequired,
  trackerInformation: PropTypes.oneOfType([
    PropTypes.shape({}),
    PropTypes.shape({
      trackerId: PropTypes.shape({
        name: PropTypes.string.isRequired,
      }),
    }),
  ]).isRequired,
  totalSize: PropTypes.number.isRequired,
  wtmData: PropTypes.shape({}).isRequired,
  starredTrackerIds: PropTypes.objectOf(PropTypes.string).isRequired,
  messageCreators: PropTypes.shape({
    sendMetrics: PropTypes.func.isRequired,
    addStarredTracker: PropTypes.func.isRequired,
    removeStarredTracker: PropTypes.func.isRequired,
  }).isRequired,
};

export default GlobalTrendsTab;
