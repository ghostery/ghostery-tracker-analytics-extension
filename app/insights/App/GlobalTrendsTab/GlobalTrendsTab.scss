@import '../../../utils/scss/color-variables.scss';

.GlobalTrendsTab {
  padding: 25px 0;
  width: 954px;
}

.GlobalTrendsTab--opaqueBackground {
  background-color: $opaque-background;
}

.GlobalTrendsTab__list {
  width: 550px;
}

.GlobalTrendsTab__listHeader {
  padding: 3px 15px 5px;
}

.GlobalTrendsTab__headerButton {
  margin: 7px;
  border: 1px solid $white;
  border-radius: 2px;
  width: 130px;
  height: 28px;
  font-size: 11px;
  cursor: pointer;

  &.active {
    background-color: $ghosty-blue;
    border: 1px solid $ghosty-blue;
  }
}

.GlobalTrendsListItem__prevalence {
  position: relative;
  left: -8px;
}

.GlobalTrendsTab__stats {
  width: 380px;
}

.GlobalTrendsTab__statDisplay {
  padding: 3px 20px 1px;
}

.GlobalTrendsTab__statText {
  margin: 0 0 15px;
  font-weight: 600;
  font-size: 24px;
}

.GlobalTrendsTab__tableHeader {
  background-color: $opaque-background;
  padding: 10px;
  width: 530px;
  height: 15px;
}

.GlobalTrendsTab__tableHeaderContainer {
  width: 500px;
}

.GlobalTrendsTab--marginRight {
  margin-right: 10px;
}
