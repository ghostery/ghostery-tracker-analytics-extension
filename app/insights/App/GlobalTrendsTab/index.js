import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import GlobalTrendsTab from './GlobalTrendsTab.jsx';
import { fetchWTMData } from '../../../store/actions/PageInfoActions';
import {
  addStarredTracker, removeStarredTracker, sendMetrics,
} from '../../../store/messages/messageCreators';

const mapStateToProps = state => ({
  totalSize: (state.PageInfoReducer.requestsBySize.hasOwnProperty('_tracker'))
    ? state.PageInfoReducer.requestsBySize._tracker.size
    : 0,
  trackerInformation: state.PageInfoReducer.trackerInformation,
  wtmData: state.PageInfoReducer.wtmData,
  starredTrackerIds: state.SettingsReducer.localSettings.starredTrackerIds,
  messageCreators: { sendMetrics, addStarredTracker, removeStarredTracker },
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ fetchWTMData }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(GlobalTrendsTab);
