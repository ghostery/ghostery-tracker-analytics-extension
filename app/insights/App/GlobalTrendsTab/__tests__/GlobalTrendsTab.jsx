import React from 'react';
import { shallow } from 'enzyme';
import path from 'path';
import fakeProps from 'react-fake-props';
import GlobalTrendsTab from '../GlobalTrendsTab';

jest.mock('../../../../../src/classes/Globals', () => ({
  BROWSER_INFO: { displayName: 'Chrome' },
}));

describe('app/insights/App/GlobalTrendsTab/', () => {
  describe('Snapshot tests with react-test-renderer', () => {
    test('GlobalTrendsTab component should render', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<GlobalTrendsTab {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('Functionality Testing', () => {
    test('handleClick state should be favorites', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const initialState = { sort: 'favorites' };
      const component = shallow(<GlobalTrendsTab {...props} />);
      component.setState({ sort: 'favorites' });

      expect(component.state()).toEqual(initialState);
    });

    test('handleClick state should be prevalence', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const initialState = { sort: 'prevalence' };
      const component = shallow(<GlobalTrendsTab {...props} />);
      component.setState({ sort: 'prevalence' });

      expect(component.state()).toEqual(initialState);
    });

    test('handleClick state should be category', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const initialState = { sort: 'category' };
      const component = shallow(<GlobalTrendsTab {...props} />);
      component.setState({ sort: 'category' });

      expect(component.state()).toEqual(initialState);
    });

    test('testing sendMetrics with favorites', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<GlobalTrendsTab {...props} />);
      component.setState({ type: 'sort_trackerInformationTab', sortType: 'favorites' });
      expect(component.find('.GlobalTrendsTab__headerButtonList').length).toBe(1);
    });

    test('testing sendMetrics with prevalence', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<GlobalTrendsTab {...props} />);
      component.setState({ type: 'sort_trackerInformationTab', sortType: 'prevalence' });
      expect(component.find('.GlobalTrendsTab__headerButtonList').length).toBe(1);
    });

    test('testing sendMetrics with category', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<GlobalTrendsTab {...props} />);
      component.setState({ type: 'sort_trackerInformationTab', sortType: 'category' });
      expect(component.find('.GlobalTrendsTab__headerButtonList').length).toBe(1);
    });

    test('handleStar should be called', () => {
      const componentPath = path.join(__dirname, '../GlobalTrendsTab.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<GlobalTrendsTab {...props} />);
      component.instance().handleStar = jest.fn();
      component.update();
      expect(component).toHaveLength(1);
    });
  });
});
