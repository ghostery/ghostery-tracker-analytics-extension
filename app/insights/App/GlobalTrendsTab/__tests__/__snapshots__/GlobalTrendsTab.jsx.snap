// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/insights/App/GlobalTrendsTab/ Snapshot tests with react-test-renderer GlobalTrendsTab component should render 1`] = `
<div
  className="GlobalTrendsTab d-flex justify-content-between align-items-start"
>
  <div
    className="GlobalTrendsTab__list GlobalTrendsTab--opaqueBackground"
  >
    <div
      className="GlobalTrendsTab__listHeader d-flex flex-column justify-content-between"
    >
      <div
        className="d-flex justify-content-start align-items-center"
      >
        <h3
          className="GlobalTrendsTab--marginRight"
        >
          Known Tracker Trends
        </h3>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              className="tooltipRight"
              placement="right"
            >
              View available statistics and trends over time for trackers on this page
            </Tooltip>
          }
          placement="auto"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <img
            alt="More info hover: View known tracker trends"
            className="infoIcon"
            height="16"
            src="/dist/images/shared/info-icon.svg"
            width="16"
          />
        </OverlayTrigger>
      </div>
      <div
        className="GlobalTrendsTab__headerButtonList d-flex justify-content-between align-items-center"
      >
        <p>
          Sort by: 
        </p>
        <OverlayTrigger
          defaultOverlayShown={false}
          delay={
            Object {
              "show": 500,
            }
          }
          overlay={
            <Tooltip
              className="tooltipLeft"
              placement="right"
            >
              Sort by favorited trackers
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <div
            className="GlobalTrendsTab__headerButton d-flex justify-content-center align-items-center active"
            onClick={[Function]}
          >
            <p>
              Favorites
            </p>
          </div>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          delay={
            Object {
              "show": 500,
            }
          }
          overlay={
            <Tooltip
              className="tooltipLeft"
              placement="right"
            >
              Sort by most common trackers
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <div
            className="GlobalTrendsTab__headerButton d-flex justify-content-center align-items-center"
            onClick={[Function]}
          >
            <p>
              Prevalence
            </p>
          </div>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          delay={
            Object {
              "show": 500,
            }
          }
          overlay={
            <Tooltip
              className="tooltipLeft"
              placement="right"
            >
              Sort by tracker type
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <div
            className="GlobalTrendsTab__headerButton d-flex justify-content-center align-items-center"
            onClick={[Function]}
          >
            <p>
              Category
            </p>
          </div>
        </OverlayTrigger>
      </div>
    </div>
    <div
      className="GlobalTrendsTab__tableHeader"
    >
      <Container
        className="GlobalTrendsTab__tableHeaderContainer"
        fluid={true}
      >
        <ForwardRef
          noGutters={false}
        >
          <Col
            className="GlobalTrendsListItem--paddingZero"
            sm={1}
          />
          <Col
            className="GlobalTrendsListItem__prevalence GlobalTrendsListItem--paddingZero"
            sm={2}
          >
            Prevalence
          </Col>
          <Col
            className="GlobalTrendsListItem--paddingZero"
            sm={5}
          >
            Tracker
          </Col>
          <Col
            className="GlobalTrendsListItem--paddingZero"
            sm={4}
          >
            Category
          </Col>
        </ForwardRef>
      </Container>
    </div>
  </div>
  <div
    className="GlobalTrendsTab__stats"
  >
    <div
      className="GlobalTrendsTab__statDisplay GlobalTrendsTab--opaqueBackground"
    >
      <div
        className="d-flex justify-content-start align-items-center"
      >
        <h3
          className="GlobalTrendsTab--marginRight"
        >
          Total Tracker Size on this Page
        </h3>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              className="tooltipLeft"
              placement="right"
            >
              Aggregate size of all trackers found on this page
            </Tooltip>
          }
          placement="auto"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <img
            alt="More info hover: Aggregate size of all trackers found on this page"
            className="infoIcon"
            height="16"
            src="/dist/images/shared/info-icon.svg"
            width="16"
          />
        </OverlayTrigger>
      </div>
      <p
        className="GlobalTrendsTab__statText"
      >
        1 B
      </p>
    </div>
  </div>
</div>
`;
